### PR TITLE
Use a bit vector in ADCE

### PR DESCRIPTION
**Pull request by s-perron**
The unordered_set in ADCE that holds all of the live instructions takes
a very long time to be destroyed.  In some shaders, it takes over 40% of
the time.

If we look at the unique ids of the live instructions, I believe they
are dense enough make a simple bit vector a good choice for to hold that
data.  When I check the density of the bit vector for larger shaders, we
are usually using less than 4 bytes per element in the vector, and
almost always less than 16.

So, in this commit, I introduce a simple bit vector class, and
use it in ADCE.

This help improve the compile time for some shaders on windows by the
40% mentioned above.

Contributes to https://github.com/KhronosGroup/SPIRV-Tools/issues/1328.

### CODE REVIEW

File: source/util/bit_vector.h [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+// limitations under the License.
+
+#ifndef LIBSPIRV_UTILS_BIT_VECTOR_H_
+#define LIBSPIRV_UTILS_BIT_VECTOR_H_
+
+#include <cassert>
+#include <iostream>
+#include <vector>
+
+#include "ilist_node.h"
```
Comment (alan-baker) [inline source/util/bit_vector.h]: Unnecessary include?

File: source/util/bit_vector.h [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef LIBSPIRV_UTILS_BIT_VECTOR_H_
+#define LIBSPIRV_UTILS_BIT_VECTOR_H_
+
+#include <cassert>
+#include <iostream>
```
Comment (alan-baker) [inline source/util/bit_vector.h]: Change this to just ostream. Even better, make it iosfwd and add a cpp that implements ReportDensity and includes ostream.

File: source/util/bit_vector.h [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+//
+// All bits default to zero, and the upper bound is 2^32-1.
+class BitVector {
+ public:
+  // Creates a bit vector contianing 0s.
+  BitVector() : bits(1024 / sizeof(BitContainer), 0) {}
+
+  // Sets the |i|th bit to 1.
+  void Set(uint32_t i) {
+    uint32_t element_index = i / sizeof(BitContainer);
```
Comment (alan-baker) [inline source/util/bit_vector.h]: sizeof returns the number of bytes. Should be dividing by 64 here to get the right vector element. The 63rd bit is in index 0, not index 7.

File: source/util/bit_vector.h [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    uint32_t bit_in_element = i % sizeof(BitContainer);
```
Comment (alan-baker) [inline source/util/bit_vector.h]: This is also wrong for 64 bits.

**Comment by s-perron** [comment]
I believe I addressed all of @alan-baker suggestions.

@dnovillo The spec does not force std::vector<bool> to be a bit vector, it could be an array of bools.  See http://en.cppreference.com/w/cpp/container/vector_bool.

**Comment by dnovillo** [comment]
On Thu, Apr 12, 2018 at 4:56 PM Steven Perron <notifications@github.com>
wrote:

> I believe I addressed all of @alan-baker <https://github.com/alan-baker>
> suggestions.
>
> @dnovillo <https://github.com/dnovillo> The spec does not force
> std::vector to be a bit vector, it could be an array of bools. See
> http://en.cppreference.com/w/cpp/container/vector_bool.
>

Hm, quite right.  I haven't run into an implementation that doesn't, but
better safe than sorry.

>

File: source/opt/aggressive_dead_code_elim_pass.h [hunk 5]
```diff
@@ -72,7 +73,10 @@ class AggressiveDCEPass : public MemPass {
 
   // Add |inst| to worklist_ and live_insts_.
   void AddToWorklist(ir::Instruction* inst) {
-    if (live_insts_.insert(inst).second) worklist_.push(inst);
+    if (!IsLive(inst)) {
```
Comment (dneto0) [inline source/opt/aggressive_dead_code_elim_pass.h]: This does two operations on the bit vector when before we only did one.

Seems we could use the abstraction of "ensure the bit is set, and tell me if it wasn't set before this operation".
Either that, or update Set to return a bool saying whether it actually changed the state of the value.  (I don't know if that would be slower to burden *every* Set operation that way.

File: source/util/bit_vector.h [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+    return (bits[element_index] &
+            (static_cast<BitContainer>(1) << bit_in_element)) != 0;
+  }
+
+  void ReportDensity(std::ostream& out);
+
+ private:
+  using BitContainer = uint64_t;
+  static const uint32_t kBitContainerSize = 64;
```
Comment (dneto0) [inline source/util/bit_vector.h]: This can be an anonymous enum:

```
enum : int {  kBitContainerSize = 64 }
```

File: source/util/bit_vector.h [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+namespace spvtools {
+namespace utils {
+
+// Implements a bit vector class.
+//
+// All bits default to zero, and the upper bound is 2^32-1.
+class BitVector {
+ public:
+  // Creates a bit vector contianing 0s.
+  BitVector() : bits(1024 / kBitContainerSize, 0) {}
```
Comment (dneto0) [inline source/util/bit_vector.h]: The 1024 is kind of magic.  Better to have

```
enum : int { kInitialNumBits  = 1024 };
```

File: source/util/bit_vector.h [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    return (bits[element_index] &
+            (static_cast<BitContainer>(1) << bit_in_element)) != 0;
+  }
+
+  void ReportDensity(std::ostream& out);
+
+ private:
+  using BitContainer = uint64_t;
```
Comment (dneto0) [inline source/util/bit_vector.h]: Minor style thing, but I'd prefer to see these using and constant definitions right at the top. (still in private visibility)

File: source/util/bit_vector.h [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+    if (element_index >= bits.size()) {
+      return false;
+    }
+
+    return (bits[element_index] &
+            (static_cast<BitContainer>(1) << bit_in_element)) != 0;
+  }
+
+  void ReportDensity(std::ostream& out);
```
Comment (dneto0) [inline source/util/bit_vector.h]: Please add a method comment. :-)

File: source/util/bit_vector.h [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+ private:
+  using BitContainer = uint64_t;
+  static const uint32_t kBitContainerSize = 64;
+  std::vector<BitContainer> bits;
```
Comment (dneto0) [inline source/util/bit_vector.h]: Member data names should end in underscore, by Google style guide.

File: test/util/bit_vector_test.cpp [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+TEST(BitVectorTest, Initialize) {
+  BitVector bvec;
+  for (int i = 1; i < 10000; i *= 2) {
+    EXPECT_FALSE(bvec.Get(i));
+  }
+}
+
+TEST(BitVectorTest, Set) {
+  BitVector bvec;
+  for (int i = 3; i < 10000; i *= 2) {
```
Comment (dneto0) [inline test/util/bit_vector_test.cpp]: It's worth adding a comment saying that 10000 > 1024, so this will excecise the resize code path.

**Comment by s-perron** [comment]
@dneto0 @dnovillo  Just finished the last changes that Dave asked for.  Please have another look.

**Comment by dneto0** [comment]
Rebased, squashed, and pushed into master as d42f65e7c

**Pull request closed without merging**